tools/clang/include/clang/SPIRV/SPIRVContext.h

```diff
@@ -133,20 +157,91 @@ class SpirvContext {
+  const MatrixType *getMatrixType(const VectorType *vecType, uint32_t vecCount);
+
+  const ImageType *getImageType(const NumericalType *, spv::Dim, bool arrayed,
+                                bool ms, ImageType::WithSampler sampled,
+                                spv::ImageFormat);
+  const SamplerType *getSamplerType() const { return samplerType; }
+  const SampledImageType *getSampledImageType(const ImageType *image);
+
+  const ArrayType *getArrayType(const SpirvType *elemType, uint32_t elemCount);
+  const RuntimeArrayType *getRunTimeArrayType(const SpirvType *elemType);
+
+  const StructType *getStructType(llvm::ArrayRef<const SpirvType *> fieldTypes,
+                                  llvm::StringRef name,
+                                  llvm::ArrayRef<llvm::StringRef> fieldNames);
+
+  const SpirvPointerType *getPointerType(const SpirvType *pointee,
+                                         spv::StorageClass);
+
+  const FunctionType *getFunctionType(const SpirvType *ret,
+                                      llvm::ArrayRef<const SpirvType *> param);
+
 private:
   /// \brief The allocator used to create SPIR-V entity objects.
   ///
   /// SPIR-V entity objects are never destructed; rather, all memory associated
   /// with the SPIR-V entity objects will be released when the SpirvContext
   /// itself is destroyed.
+  ///
+  /// This field must appear the first since it will be used to allocate object
+  /// for the other fields.
   mutable llvm::BumpPtrAllocator allocator;
+
+  // Unique types
+
+  const VoidType *voidType;
+  const BoolType *boolType;
+
+  // The type at index i is for bitwidth 2^i. So max bitwidth supported
+  // is 2^6 = 64. Index 0/1/2/3 is not used right now.
+  std::array<const IntegerType *, 7> sintTypes;
+  std::array<const IntegerType *, 7> uintTypes;
+  std::array<const FloatType *, 7> floatTypes;
+
+  using VectorTypeArray = std::array<const VectorType *, 5>;
+  using MatrixTypeArray = std::array<const MatrixType *, 5>;
+  using CountToArrayMap = llvm::DenseMap<uint32_t, const ArrayType *>;
+  using SCToPtrTyMap =
+      llvm::DenseMap<spv::StorageClass, const SpirvPointerType *,
+                     StorageClassDenseMapInfo>;
+
+  // Vector/matrix types for each possible element count.
+  // Type at index is for vector of index components. Index 0/1 is unused.
+
+  llvm::DenseMap<const ScalarType *, VectorTypeArray> vecTypes;
+  llvm::DenseMap<const VectorType *, MatrixTypeArray> matTypes;
+
+  llvm::SmallVector<const ImageType *, 8> imageTypes;
+  const SamplerType *samplerType;
+  llvm::DenseMap<const ImageType *, const SampledImageType *> sampledImageTypes;
+
+  llvm::DenseMap<const SpirvType *, CountToArrayMap> arrayTypes;
+  llvm::DenseMap<const SpirvType *, const RuntimeArrayType *> runtimeArrayTypes;
+
+  llvm::SmallVector<const StructType *, 8> structTypes;
+
+  llvm::DenseMap<const SpirvType *, SCToPtrTyMap> pointerTypes;
```

Comment:
I wonder if it would be better to use
```cpp
using PointerTypeArray = std::array<const SpirvPointerType *, 12>;
llvm::DenseMap<const SpirvType *, PointerTypeArray> pointerTypes;
```

And use the ```static_cast<uint32_t>(sc)``` to index into the ```PointerTypeArray```.
That would change 2 map lookups to 1 map lookup and 1 array indexing.

Probably not a big difference. Up to you.

Reply:
It's debatable. For int/float/vector/matrix types, I used `std::array` because I think typically in a real-world shader they are all likely to be instantiated.  Even if not, we are not wasting a lot of memory.

For pointers, typically we only have 2-3 storage class for each pointee type, and we will have a lot of pointer types. So using arrays are likely waste a lot of space, and using maps with only one or two keys won't be that bad I think.

Reply:
makes sense